pa11ycrawler/pipelines.py

```diff
@@ -20,89 +21,150 @@ class DuplicatesPipeline(object):
     def __init__(self):
         self.urls_seen = set()
 
-    def process_item(self, item, _spider):
+    def clean_url(self, url):
+        """
+        A `next` query parameter indicates where the user should be
+        redirected to next -- it doesn't change the content of the page.
+        Two URLs that only differ by their `next` query paramenter should
+        be considered the same.
+        """
+        return URLObject(url).del_query_param("next")
+
+    def process_item(self, item, spider):  # pylint: disable=unused-argument
         """
         Stops processing item if we've already seen this URL before.
         """
-        if item['url'] in self.urls_seen:
-            raise DropItem("[Skipping] Duplicate url found: {}".format(item['url']))
+        url = self.clean_url(item["url"])
+        if url in self.urls_seen:
+            raise DropItem("Dropping duplicate url {url}".format(url=item["url"]))
         else:
-            self.urls_seen.add(item['url'])
+            self.urls_seen.add(url)
             return item
 
 
-class Pa11yPipeline(object):
+class DropDRFPipeline(object):
     """
-    Runs the Pa11y CLI against `item['url']`, using the same headers used by
-    Scrapy. Stores the results in `item['results']` to be processed by
-    `spider.a11y_reporter`.
+    Drop pages that are generated from Django Rest Framework (DRF), so that
+    they don't get processed by pa11y later in the pipeline.
     """
-    def __init__(self):
-        self.pa11y_settings = None
+    def process_item(self, item, spider):  # pylint: disable=unused-argument
+        "Check for DRF urls."
+        url = URLObject(item["url"])
+        if url.path.startswith("/api/"):
+            raise DropItem("Dropping DRF url {url}".format(url=url))
+        else:
+            return item
 
-    def get_pa11y_settings(self, spider):
-        """
-        Sets `self.pa11y_settings` to the list of flags that should
-        be passed to the Pa11y CLI.
-        """
-        settings = []
-        for key, value in spider.settings.attributes.iteritems():
-            if is_pa11y_setting(key) and value.value:
-                flag_name = key.replace('_', '-').split('PA11Y-')[1].lower()
-                flag = '--{}="{}"'.format(flag_name, value.value)
-                settings.append(flag)
 
-        self.pa11y_settings = settings
-        return settings
+class Pa11yPipeline(object):
+    """
+    Runs the Pa11y CLI against `item['url']`, using the same request headers
+    used by Scrapy.
+    """
+    pa11y_path = "node_modules/.bin/pa11y"
```

Comment:
I added a commit to make the `Pa11yPipeline` check to see if `pa11y` is installed when the pipeline is instantiated. If not, it raises a `scrapy.exceptions.NotConfigured` exception.